engine/kv_engine.cpp

```diff
@@ -1739,6 +1753,102 @@ void KVEngine::backgroundDramCleaner() {
   }
 }
 
+Status KVEngine::destroyExpiredList(
+    List* list, std::deque<PendingFreeSpaceEntries>* list_space_entries) {
+  PendingFreeSpaceEntries space_entries;
+  space_entries.release_time = version_controller_.GetCurrentTimestamp();
+  Status s = listDestroy(
+      list,
+      [&](void* addr, TimeStampType ts) {
+        DataEntry* data_entry = static_cast<DataEntry*>(addr);
+        space_entries.entries.emplace_back(
+            SpaceEntry{pmem_allocator_->addr2offset_checked(addr),
+                       data_entry->header.record_size});
+        space_entries.release_time = std::max(space_entries.release_time, ts);
+      },
+      false);
+  if (s == Status::Ok && !space_entries.entries.empty()) {
+    list_space_entries->emplace_back(std::move(space_entries));
+    if (old_records_cleaner_.TryFreePendingSpace(list_space_entries->front())) {
+      list_space_entries->pop_front();
+    }
+  }
+  return s;
+}
+
+void KVEngine::releaseCollections() {
+  for (auto list : lists_) {
+    delete list;
```

Comment:
Don't modify members in std::set. That's invalid.
You should get them by begin() one by one, erase it from std::set and then delete it.